dali/kernels/signal/moving_mean_square.cc

```diff
@@ -0,0 +1,141 @@
+// Copyright (c) 2020, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "dali/kernels/signal/moving_mean_square.h"
+
+namespace dali {
+namespace kernels {
+namespace signal {
+
+namespace {
+
+/**
+ * Accurate type is such type, that doesn't require
+ * reset interval for maintaining numeric precision
+ */
+template<typename T>
+struct is_accurate {
+  static constexpr bool value = std::is_integral<T>::value && sizeof(T) <= 2;
+};
+
+
+/**
+ * Type of the accumulator:
+ * In case calculation is performed on floating-point,
+ * it requires reset_interval to keep numeric accuracy.
+ */
+template<typename T>
+struct accumulator_type {
+  using type = std::conditional_t<is_accurate<T>::value, int64_t, float>;
+};
+
+template<typename T>
+using acc_t = typename accumulator_type<T>::type;
+
+
+template<typename T>
+acc_t<T> Square(const T &val) {
+  acc_t<T> res = val;
+  return res * res;
+}
+
+
+template<typename T>
+acc_t<T> CalcSumSquared(span<const T> values) {
+  acc_t<T> sumsq = 0;
+  for (const auto &val : values) {
+    sumsq += Square(val);
+  }
+  return sumsq;
+}
+
+}  // namespace
+
+template<typename T>
+MovingMeanSquareCpu<T>::~MovingMeanSquareCpu() = default;
+
+
+template<typename T>
+KernelRequirements
+MovingMeanSquareCpu<T>::Setup(KernelContext &context, const InTensorCPU<T, 1> &in,
+                              const MovingMeanSquareArgs &args) {
+  DALI_ENFORCE(args.window_size <= in.num_elements(),
+               make_string("window_size can't be bigger than input buffer. Received: window_size=",
+                           args.window_size, ", input_size=", in.num_elements()));
+  KernelRequirements req;
+  TensorShape<> out_shape = {in.shape[0] - args.window_size + 1};
+  req.output_shapes = {TensorListShape<>({out_shape})};
+  return req;
+}
+
+
+template<typename T>
+void
+CalcMovingMeanSquareWithInterval(span<float> out, span<const T> in, int length, float mean_factor,
```

Comment:
Suggestion:
move reset_interval to the end and give it a default value of -1 and handle it here. It makes this function self-contained and you won't need two functions.